plugins/ingest/src/main/java/org/elasticsearch/ingest/processor/mutate/MutateProcessor.java

```diff
@@ -0,0 +1,312 @@
+    }
+
+    private void doLowercase(Data data) {
+        if (lowercase == null) { return; }
+
+        for(String field : lowercase) {
+            Object val = data.getProperty(field);
+            if (val instanceof String) {
+                data.addField(field, ((String) val).toLowerCase(Locale.US));
+            } else {
+                throw new IllegalArgumentException("Cannot lowercase field:" + field + " with type: " + val.getClass());
+            }
+        }
+    }
+
+    public static class Builder implements Processor.Builder {
+
+        private Map<String, Object> update;
+        private Map<String, String> rename;
+        private Map<String, String> convert;
+        private Map<String, String> split;
+        private Map<String, List<String>> gsub;
+        private Map<String, String> join;
+        private List<String> remove;
+        private List<String> strip;
+        private List<String> uppercase;
+        private List<String> lowercase;
+
+        public void setUpdate(Map<String, Object> update) {
+            this.update = update;
+        }
+
+        public void setRename(Map<String, String> rename) {
+            this.rename = rename;
+        }
+
+        public void setConvert(Map<String, String> convert) {
+            this.convert = convert;
+        }
+
+        public void setSplit(Map<String, String> split) {
+            this.split = split;
+        }
+
+        public void setGsub(Map<String, List<String>> gsub) {
+            this.gsub = gsub;
+        }
+
+        public void setJoin(Map<String, String> join) {
+            this.join = join;
+        }
+
+        public void setRemove(List<String> remove) {
+            this.remove = remove;
+        }
+
+        public void setStrip(List<String> strip) {
+            this.strip = strip;
+        }
+
+        public void setUppercase(List<String> uppercase) {
+            this.uppercase = uppercase;
+        }
+
+        public void setLowercase(List<String> lowercase) {
+            this.lowercase = lowercase;
+        }
+
+        @SuppressWarnings("unchecked")
+        public void fromMap(Map<String, Object> config) {
+            this.update = (Map<String, Object>) config.get("update");
+            this.rename = (Map<String, String>) config.get("rename");
+            this.convert = (Map<String, String>) config.get("convert");
+            this.split = (Map<String, String>) config.get("split");
+            this.gsub = (Map<String, List<String>>) config.get("gsub");
+            this.join = (Map<String, String>) config.get("join");
+            this.remove = (List<String>) config.get("remove");
+            this.strip = (List<String>) config.get("strip");
+            this.uppercase = (List<String>) config.get("uppercase");
+            this.lowercase = (List<String>) config.get("lowercase");
+        }
+
+        @Override
+        public Processor build() {
+            return new MutateProcessor(update, rename, convert, split, gsub,
```

Comment:
maybe wrap all these maps in `Collections#unmodifiableMap(...)`? To enforce that no changes can be made.